Matrix question

Dhruwang opened the pull request:
<!-- We require pull request titles to follow the Conventional Commits specification ( https://www.conventionalcommits.org/en/v1.0.0/#summary ). Please make sure your title follow these conventions -->

## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

Fixes [1185](https://linear.app/formbricks/issue/FOR-1185/question-type-matrix-table)

Preview:
<img width="583" alt="Screenshot 2024-03-18 at 11 57 39 AM" src="https://github.com/formbricks/formbricks/assets/67850763/4e16243a-e503-4d37-ad8e-7f872e322e81">


Matrix question summary:
<img width="1256" alt="Screenshot 2024-03-18 at 11 59 05 AM" src="https://github.com/formbricks/formbricks/assets/67850763/b3b3e38b-3e68-4d26-8833-8e705d7c75ba">

Respones Card
<img width="378" alt="Screenshot 2024-03-18 at 12 00 19 PM" src="https://github.com/formbricks/formbricks/assets/67850763/f58e6a42-d025-474f-b6e2-6ccfdc5b424a">

Embed survey
<img width="1038" alt="Screenshot 2024-03-18 at 11 59 30 AM" src="https://github.com/formbricks/formbricks/assets/67850763/8170540a-1266-482e-90aa-56a8b8a26a8c">

Response received email
<img width="650" alt="Screenshot 2024-03-18 at 12 02 03 PM" src="https://github.com/formbricks/formbricks/assets/67850763/4710fa26-582c-449e-91ea-50e7cbc36d51">

CSV download 
<img width="378" alt="Screenshot 2024-03-18 at 11 59 59 AM" src="https://github.com/formbricks/formbricks/assets/67850763/5ba83947-1ac1-405e-a292-78eac1dcfbaf">

Integration
<img width="532" alt="Screenshot 2024-03-18 at 12 21 58 PM" src="https://github.com/formbricks/formbricks/assets/67850763/1b247af7-c956-4a07-b0ea-dc112cde9d9f">


<!-- Please provide a screenshots or a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

Sections provided above
## Checklist

<!-- We're starting to get more and more contributions. Please help us making this efficient for all of us and go through this checklist. Please tick off what you did  -->

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [x] Updated the Formbricks Docs if changes were necessary

ShubhamPalriwala commented on apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/components/MatrixQuestionForm.tsx:
```diff
@@ -0,0 +1,180 @@
+  localSurvey: TSurvey;
+  question: TSurveyMatrixQuestion;
+  questionIdx: number;
+  updateQuestion: (questionIdx: number, updatedAttributes: any) => void;
```
Can we typecast the `updatedAttributes` away from any please?

Dhruwang replied:
Yes this was the case with every other questions as well, changed them too 😊

ShubhamPalriwala commented on packages/lib/response/util.ts:
```diff
@@ -971,6 +981,58 @@ export const getQuestionWiseSummary = (
+        let totalResponseCount = 0;
+
+        // Initialize count object
+        const countMap: Record<string, string> = rows.reduce((acc, row) => {
+          acc[row] = columns.reduce((colAcc, col) => {
+            colAcc[col] = 0;
+            return colAcc;
+          }, {});
+          return acc;
+        }, {});
```
Can we have better naming variables for more readability here?

Dhruwang replied:
We already use choiceCountMap in multiChoice summary so to keep it consistent, its named this way and also its a map of counting and hence is being named countMap 😊

ShubhamPalriwala commented on packages/lib/responses.ts:
```diff
@@ -39,3 +39,21 @@ export const getQuestionResponseMapping = (
   return questionResponseMapping;
 };
+
+export const processResponseData = (
+  responseData: string | number | string[] | Record<string, string>
+): string => {
+  if (!responseData) return "";
+  if (typeof responseData === "string") {
+    return responseData;
+  } else if (typeof responseData === "number") {
+    return responseData.toString();
+  } else if (Array.isArray(responseData)) {
+    return responseData.join("; ");
+  } else {
+    const formattedString = Object.entries(responseData)
+      .map(([key, value]) => `${key}: ${value}`)
+      .join("\n");
+    return formattedString;
+  }
+};
```
Can we convert this to a `switch`? Since it has so many conditional statements & returns based on that, looks like a better option?

ShubhamPalriwala reviewed (commented):
Hey @Dhruwang this looks amazing! 🔥 Just saw the functionality & the use cases are promising. Well built even on the responsiveness! Once the Jumping is ready, it'd be even nice to test 😉 

I ran into a few small bugs during testing, could you please take a look at them?
Bugs:
- Can we follow the same order of the rows that we do during question creation:
![image](https://github.com/formbricks/formbricks/assets/55556994/62326195-2623-4adc-babd-e3af86dea98a)
In the response it currently looks like:
![image](https://github.com/formbricks/formbricks/assets/55556994/d3eb6454-dde7-4def-b632-956977c6dc16)


- I cannot deselect an option from a row:
![image](https://github.com/formbricks/formbricks/assets/55556994/ddb314bf-7a56-44a3-a817-55aedfe5102a)

- In the survey editor, when adding a column, on pressing enter (expecting to enter a new column), i enter a new row:
![image](https://github.com/formbricks/formbricks/assets/55556994/1c9d887a-ea04-4eb0-b0c0-4f1c99cb9ada)

- On adding a realllyyy long column name, the UI breaks
![image](https://github.com/formbricks/formbricks/assets/55556994/df19f797-2b49-4cef-8944-1a33a3569532)
![image](https://github.com/formbricks/formbricks/assets/55556994/0edd3eda-1143-48a0-b28b-1e0ac2b2f94b)
- Same for rows as well: 
![image](https://github.com/formbricks/formbricks/assets/55556994/9c454769-1456-41c8-9fa6-6f9402839c33)
Can we use tooltips here as mentioned in the issue & limit with max heights & widhts?

- Even if the question is optional, it still does not let me move forward without filling the grid in:
![image](https://github.com/formbricks/formbricks/assets/55556994/1e3d64d0-2bb0-42cc-b55e-d5d07790c79b)

Dhruwang commented on apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/components/MatrixQuestionSummary.tsx:
```diff
@@ -0,0 +1,112 @@
+      </Tooltip>
+    </TooltipProvider>
+  );
+};
```
Yes we do have a `ToopTipRenderer` component in `packages/ui` , so have reused that 😊

Dhruwang commented on packages/surveys/src/components/general/QuestionConditional.tsx:
```diff
@@ -46,7 +47,7 @@ export default function QuestionConditional({
     <OpenTextQuestion
       question={question}
-      value={value}
+      value={typeof value === "string" ? value : ""}
```
Wasn't able to come up with a good solution here, do you have anything on mind?🤔

ShubhamPalriwala replied:
Try asking GPT! It should help I think!

Dhruwang commented:
Hey @ShubhamPalriwala , have updated the PR with mentioned changes 😊

Regarding the issue of the UI breaking with lengthy row and column names, we've chosen to implement tooltips exclusively on the summary page. However, for the question page, our approach is to display the entire label and offer users the necessary scrolling options to view these labels in full.

ShubhamPalriwala commented on packages/ui/QuestionFormInput/utils.ts:
```diff
@@ -26,6 +28,20 @@ export const getChoiceLabel = (
+  surveyLanguageCodes: string[],
+  type: "row" | "column"
+): TI18nString => {
+  const matrixQuestion = question as TSurveyMatrixQuestion;
+  if (type === "row") {
+    return matrixQuestion.rows[idx] || createI18nString("", surveyLanguageCodes);
+  } else {
+    return matrixQuestion.columns[idx] || createI18nString("", surveyLanguageCodes);
+  }
```
Can we rewrite this to something like below for cleaner code:
```ts
const matrixQuestion = question as TSurveyMatrixQuestion;
  const labels = type === "row" ? matrixQuestion.rows : matrixQuestion.columns;
  return labels[idx] || createI18nString("", surveyLanguageCodes);
```

ShubhamPalriwala commented on packages/ui/QuestionFormInput/index.tsx:
```diff
@@ -459,7 +478,7 @@ export const QuestionFormInput = ({
         )}
       </div>
-      {selectedLanguageCode !== "default" && value && value["default"] && (
+      {selectedLanguageCode !== "default" && value && (value["default"] || value["default"] === "") && (
```
Can this condition also be simplified if possible for readability? 
Maybe GPT says: `selectedLanguageCode !== "default" && value?.["default"] ?? "" !== "";` but can you test & see if it works and no corner case is missed?

ShubhamPalriwala commented on apps/web/playwright/utils/mock.ts:
```diff
@@ -21,17 +21,17 @@ export const users = {
-      email: "survey1@formbricks.com",
+      email: "survey28@formbricks.com",
       password: "Y1I*EpURUSb32j5XijP",
     },
     {
       name: "Survey User 2",
-      email: "survey2@formbricks.com",
+      email: "survey29@formbricks.com",
       password: "G73*Gjif22F4JKM1pA",
     },
     {
       name: "Survey User 3",
-      email: "survey3@formbricks.com",
+      email: "survey43@formbricks.com",
```
Can we revert this back please? :)

ShubhamPalriwala reviewed (commented):
Great work! Looks pretty solid to me! Cannot face any of the previous bugs anymore 💪🏼 
Just one small fix: 
![Screenshot_2024-03-22-12-21-06_5760x1080](https://github.com/formbricks/formbricks/assets/55556994/efbec0b3-494d-4dad-8335-9e4949b8f73e)
In the above, I see a cursor pointer on the heatmap when hovered? is it supposed to do something on click? is this behaviour intended?

Apart from these, just 2-3 small code changes, rest all looks good from my end 🚀

ShubhamPalriwala reviewed (approved):
LGTM ✅ 🚀

jobenjada commented:
hey! here's what I noticed (if unrelated to this PR pls fix in a separat PR):

1. default NPS quesiton:

<img width="512" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/c66492b9-e19b-4ee3-9310-3b9ecbfbbf8b">

-----

2. CTA question doesn't have a text by default

<img width="520" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/383f83ae-ff18-4297-9907-a1ed16c10f46">

-------

3. Unusual red frame for file upload
<img width="541" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/b2e8cf81-6291-4378-979c-3b3a13f88269">

------

4. It's kind of hard to hit the radio button. I know there are a few CSS tricks to make it easier to click (e.g. adding an invisible ring or similar. Pls choose a solution which works stable on all browsers and devices:

<img width="135" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/95fc66b5-e0ba-4691-a497-5cfec6e81500">

--------

5. Make use of new css classes so that the question can be styled with new styling (colors and roundness)
<img width="515" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/cbc69da4-f54f-46f3-9d4e-50e7f65d3aad">
<img width="461" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/07e3cf0a-3be3-4258-8b00-cc2a545facaa">

-------
5.b replace color with styled color

<img width="473" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/74769af1-c47a-4dec-94f9-edbe78968f5f">

-------

6. the spacing is a bit off, reduce a bit
<img width="466" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/a9033be6-cb58-496f-99ba-677af21fb28a">

-------

7. contrast is weak, not sure which color this text is but should be slate-950 and reduce the darkest teal so that it does not get as dark and the contrast is better:
<img width="426" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/f70bad59-0333-4907-8aab-a26536e3306d">

------

8. replace with filled star  (not only outline). It's ok to custom SVG component for this this lucide doesnt provide
<img width="316" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/03defee3-f3b0-4b6c-8c55-a78d843ab88d">

-----

9. Remove space before the ":"
<img width="132" alt="image" src="https://github.com/formbricks/formbricks/assets/72809645/25f287fe-ad1e-4ae8-9f59-61da5ef9139b">

-----

The rest looks good, thanks! :))

Dhruwang commented:
Hey @jobenjada , have fixed the mentioned issues
Issue 1.2,3,5b and 8 seems unrelated to this PR (tested on prod) so will create a separate PR for it 😊


Fixed issue - 4

https://github.com/formbricks/formbricks/assets/67850763/41cb7220-b976-4839-91b6-06b99dff9437



Fixed issue - 5
![Screenshot 2024-03-29 at 10 30 14 AM](https://github.com/formbricks/formbricks/assets/67850763/b5506dcd-6be1-45e8-8882-83afba7c7042)

Fixed issue - 7 

![Screenshot 2024-03-29 at 10 32 33 AM](https://github.com/formbricks/formbricks/assets/67850763/ced52d08-f3f9-4ced-8acd-440010bdb438)

You can also adjust the value for it in function `getOpacityLevel` in `MatrixQuestionSummary.tsx`, where you can change value of 15 between 0 to 25, where moving towards 25 makes it darker
![Screenshot 2024-03-29 at 10 42 51 AM](https://github.com/formbricks/formbricks/assets/67850763/afc235d8-3993-4175-995b-52a328f50a85)



Fixed issue - 6
![Screenshot 2024-03-29 at 10 32 08 AM](https://github.com/formbricks/formbricks/assets/67850763/3d06bcef-2422-408f-af82-ec75087d6a18)

Fixed issue - 9
![Screenshot 2024-03-29 at 10 39 46 AM](https://github.com/formbricks/formbricks/assets/67850763/725ab7cf-b624-4b78-9caa-e6471914e7c1)

jobenjada commented:
@mattinannt LGTM

jobenjada reviewed (approved):
LGTM
